### ISSUE TITLE
Exception: 'TokenizersWrapper' object has no attribute 'bos_token_id'

### ISSUE THREAD

**Issue by ChunkDunkirk** [bmaltais/kohya_ss#3173]
I've been beating my head against this for a while trying to train SDXL LoRAs. I've used more or less this exact setup with these settings before with success, so I'm not sure what I'm doing wrong. I am using ComfyUI Portable for Windows, with a virtual python environment. Trying to use either the full node or the "Easy Train" plugin, When I use the full Kohya SS node within ComfyUI, I am constantly returned with the error "Exception: 'TokenizersWrapper' object has no attribute 'bos_token_id'". I've run pip to uninstall and reinstall tokenizers, made sure tokenizers were installed and recognized by the system, I've run pip requirements.txt, I've tried starting with a fresh comfyui install and virtual environment, made sure the environment was activated, and so on and so on. No luck. I've run ComfyUI within the venv, and standard image generation and whatnot works just fine. I can provide more info and context as needed, if I'm forgetting anything off the top of my head. Any help I can get at this point, would be greatly appreciated. Below is the full log from the process:

> =======================train_config=======================
>     {
>     "pretrained_model_name_or_path": "D:\\Games\\ComfyUI_windows_portable\\ComfyUI\\models\\checkpoints\\trueAmateurFeelingXL_v2.safetensors",
>     "max_train_steps": "0",
>     "xformers": false,
>     "sdpa": false,
>     "fp8_base": false,
>     "mixed_precision": "fp16",
>     "cache_latents": true,
>     "cache_latents_to_disk": true,
>     "network_dim": "16",
>     "network_alpha": "8",
>     "network_module": "networks.lora",
>     "network_train_unet_only": true,
>     "learning_rate": "0.00001",
>     "lr_scheduler": "cosine",
>     "optimizer_type": "AdamW",
>     "save_every_n_epochs": "1",
>     "shuffle_caption": false,
>     "lr_warmup_steps": "0",
>     "save_precision": "fp16",
>     "lr_scheduler_num_cycles": "1",
>     "persistent_data_loader_workers": true,
>     "no_metadata": true,
>     "noise_offset": "0.1",
>     "output_dir": "D:\\Games\\ComfyUI_windows_portable\\ComfyUI\\output\\FL_train_workspaces\\DSS1\\output",
>     "output_name": "DSS1_lora_sdxl_20250410_130023",
>     "no_half_vae": true,
>     "lowram": true,
>     "dataset_config": "D:\\Games\\ComfyUI_windows_portable\\ComfyUI\\output\\FL_train_workspaces\\DSS1\\dataset.json",
>     "max_train_epochs": "10",
>     "gradient_accumulation_steps": "1",
>     "gradient_checkpointing": false
> }
> 
> =======================other_config=======================
>     {
>     "sample_prompt": ""
> }
> 
> train_func = run_lora_sdxl
> D:\Games\ComfyUI_windows_portable\ComfyUI\custom_nodes\comfyui_fl-trainer\kohya_ss_lora\library\custom_train_functions.py:163: SyntaxWarning: invalid escape sequence '\('
>   """
> D:\Games\ComfyUI_windows_portable\ComfyUI\custom_nodes\comfyui_fl-trainer\kohya_ss_lora\library\lpw_stable_diffusion.py:64: SyntaxWarning: invalid escape sequence '\('
>   """
> D:\Games\ComfyUI_windows_portable\ComfyUI\custom_nodes\comfyui_fl-trainer\kohya_ss_lora\library\sdxl_lpw_stable_diffusion.py:68: SyntaxWarning: invalid escape sequence '\('
>   """
> 2025-04-10 13:00:29 INFO     Loading dataset config from                                            train_network.py:161
>                              D:\Games\ComfyUI_windows_portable\ComfyUI\output\FL_train_workspaces\D
>                              SS1\dataset.json
>                     INFO     prepare images.                                                          train_util.py:1572
>                     INFO     found directory                                                          train_util.py:1519
>                              D:\Games\ComfyUI_windows_portable\ComfyUI\output\FL_train_workspaces\DSS
>                              1\train_images contains 10 image files
>                     INFO     200 train images with repeating.                                         train_util.py:1613
>                     INFO     0 reg images.                                                            train_util.py:1616
>                     WARNING  no regularization images / 正則化画像が見つかりませんでした              train_util.py:1621
>                     INFO     [Dataset 0]                                                              config_util.py:565
>                                batch_size: 1
>                                resolution: (512, 512)
>                                enable_bucket: True
>                                network_multiplier: 1.0
>                                min_bucket_reso: 256
>                                max_bucket_reso: 1024
>                                bucket_reso_steps: 64
>                                bucket_no_upscale: False
> 
>                                [Subset 0 of Dataset 0]
>                                  image_dir:
>                              "D:\Games\ComfyUI_windows_portable\ComfyUI\output\FL_train_workspaces\DS
>                              S1\train_images"
>                                  image_count: 10
>                                  num_repeats: 20
>                                  shuffle_caption: False
>                                  keep_tokens: 0
>                                  keep_tokens_separator:
>                                  secondary_separator: None
>                                  enable_wildcard: False
>                                  caption_dropout_rate: 0.0
>                                  caption_dropout_every_n_epoches: 0
>                                  caption_tag_dropout_rate: 0.0
>                                  caption_prefix: None
>                                  caption_suffix: None
>                                  color_aug: False
>                                  flip_aug: False
>                                  face_crop_aug_range: None
>                                  random_crop: False
>                                  token_warmup_min: 1,
>                                  token_warmup_step: 0,
>                                  is_reg: False
>                                  class_tokens: None
>                                  caption_extension: .txt
> 
> 
>                     INFO     [Dataset 0]                                                              config_util.py:571
>                     INFO     loading image sizes.                                                      train_util.py:853
> 100%|█████████████████████████████████████████████████████████████████████████████████| 10/10 [00:00<00:00, 294.12it/s]
>                     INFO     make buckets                                                              train_util.py:859
>                     INFO     number of images (including repeats) /                                    train_util.py:905
>                              各bucketの画像枚数（繰り返し回数を含む）
>                     INFO     bucket 0: resolution (512, 512), count: 200                               train_util.py:910
>                     INFO     mean ar error (without repeats): 0.0                                      train_util.py:915
>                     INFO     preparing accelerator                                                  train_network.py:225
> accelerator device: cuda
>                     INFO     loading model for process 0/1                                         sdxl_train_util.py:30
> load Diffusers pretrained models: D:\Games\ComfyUI_windows_portable\ComfyUI\models\checkpoints\trueAmateurFeelingXL_v2.safetensors, variant=fp16
> _snapshot_download: stabilityai/stable-diffusion-xl-base-1.0
> Loading pipeline components...: 100%|████████████████████████████████████████████████████| 7/7 [00:06<00:00,  1.14it/s]
> U-Net converted to original U-Net
> import network module: networks.lora
> 2025-04-10 13:00:43 INFO     [Dataset 0]                                                              train_util.py:2079
>                     INFO     caching latents.                                                          train_util.py:974
>                     INFO     checking cache validity...                                                train_util.py:984
> 100%|██████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:00<?, ?it/s]
>                     INFO     caching latents...                                                       train_util.py:1021
> 100%|██████████████████████████████████████████████████████████████████████████████████| 10/10 [03:12<00:00, 19.25s/it]
> 2025-04-10 13:03:56 INFO     create LoRA network. base dim (rank): 16, alpha: 8.0                            lora.py:810
>                     INFO     neuron dropout: p=None, rank dropout: p=None, module dropout: p=None            lora.py:811
>                     INFO     create LoRA for Text Encoder 1:                                                 lora.py:902
>                     INFO     create LoRA for Text Encoder 2:                                                 lora.py:902
>                     INFO     create LoRA for Text Encoder: 88 modules.                                       lora.py:910
>                     INFO     create LoRA for U-Net: 722 modules.                                             lora.py:918
>                     INFO     enable LoRA for U-Net                                                           lora.py:966
> prepare optimizer, data loader etc.
>                     INFO     use AdamW optimizer | {}                                                 train_util.py:4087
> override steps. steps for 10 epochs is / 指定エポックまでのステップ数: 2000
> running training / 学習開始
>   num train images * repeats / 学習画像の数×繰り返し回数: 200
>   num reg images / 正則化画像の数: 0
>   num batches per epoch / 1epochのバッチ数: 200
>   num epochs / epoch数: 10
>   batch size per device / バッチサイズ: 1
>   gradient accumulation steps / 勾配を合計するステップ数 = 1
>   total optimization steps / 学習ステップ数: 2000
> steps:   0%|                                                                                  | 0/2000 [00:00<?, ?it/s]sample_images: height = 512, width = 512
> 2025-04-10 13:04:03 INFO     prompt:                                                                  train_util.py:5284
>                     INFO     negative_prompt:                                                         train_util.py:5285
>                     INFO     height: 512                                                              train_util.py:5286
>                     INFO     width: 512                                                               train_util.py:5287
>                     INFO     sample_steps: 20                                                         train_util.py:5288
>                     INFO     scale: 5.0                                                               train_util.py:5289
>                     INFO     sample_sampler: euler_a                                                  train_util.py:5290
>                     INFO     seed: 0                                                                  train_util.py:5292
> Exception: 'TokenizersWrapper' object has no attribute 'bos_token_id'
> Press Enter to continue...

**Comment by ChunkDunkirk** [comment]
As an update to this, I realized that there were **two** requirements.txt files, one within the root Kohya-SS folder, and one in the sub-folder sd-scripts. I activated my venv and ran pip on the requirements.txt file in sd-scripts, now I'm returning this error, one I haven't seen before. "Exception: module 'library.sdxl_train_util' has no attribute 'SdxlStableDiffusionLongPromptWeightingPipeline'":

> =======================train_config=======================
>     {
>     "pretrained_model_name_or_path": "D:\\Games\\ComfyUI_windows_portable\\ComfyUI\\models\\checkpoints\\trueAmateurFeelingXL_v2.safetensors",
>     "max_train_steps": "1000000",
>     "xformers": false,
>     "sdpa": false,
>     "fp8_base": false,
>     "mixed_precision": "fp16",
>     "cache_latents": true,
>     "cache_latents_to_disk": true,
>     "network_dim": "64",
>     "network_alpha": "32",
>     "network_module": "networks.lora",
>     "network_train_unet_only": false,
>     "learning_rate": "1.0000000000000004e-05",
>     "lr_scheduler": "constant",
>     "optimizer_type": "AdaFactor",
>     "save_every_n_epochs": "1",
>     "shuffle_caption": true,
>     "lr_warmup_steps": "0",
>     "save_precision": "fp16",
>     "lr_scheduler_num_cycles": "1",
>     "persistent_data_loader_workers": false,
>     "no_metadata": false,
>     "noise_offset": "0.05",
>     "output_dir": "D:\\Games\\ComfyUI_windows_portable\\ComfyUI\\output\\FL_train_workspaces\\my_lora\\output",
>     "output_name": "my_lora_lora_sdxl_20250410_144345",
>     "no_half_vae": true,
>     "lowram": true,
>     "dataset_config": "D:\\Games\\ComfyUI_windows_portable\\ComfyUI\\output\\FL_train_workspaces\\my_lora\\dataset.json",
>     "max_train_epochs": "10",
>     "gradient_accumulation_steps": "1",
>     "gradient_checkpointing": false,
>     "unet_lr": "1.0000000000000004e-05",
>     "text_encoder_lr": "5.000000000000002e-06"
> }
> 
> =======================other_config=======================
>     {
>     "sample_prompt": "a realistic portrait of a woman with dark brown hair"
> }
> 
> train_func = run_lora_sdxl
> Exception: module 'library.sdxl_train_util' has no attribute 'SdxlStableDiffusionLongPromptWeightingPipeline'
> Press Enter to continue...
